io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -619,42 +574,59 @@ exploreSimTraceST optsf main k =
       ppIOSimThreadId tid ++ (case lab of Nothing -> ""
                                           Just l  -> " ("++l++")")
 
+    -- cache of explored schedules
+    cache :: IORef (Set ScheduleControl)
+    cache = unsafePerformIO cacheIO
+
     -- insert a schedule into the cache
-    cachedST :: STRef s (Set ScheduleControl) -> ScheduleControl -> ST s (Maybe ScheduleControl)
-    cachedST cacheRef a = do
-      set <- readSTRef cacheRef
-      writeSTRef cacheRef (Set.insert a set)
-      return $ if Set.member a set
-               then Nothing
-               else Just a
+    cached :: ScheduleControl -> Bool
+    cached = unsafePerformIO . cachedIO
+
+    -- compute cache size; it's a function to make sure that `GHC` does not
+    -- inline it (and share the same thunk).
+    cacheSize :: () -> Int
+    cacheSize = unsafePerformIO . cacheSizeIO
 
     --
-    -- Caching in ST monad
+    -- Caching in IO monad
     --
 
     -- It is possible for the same control to be generated several times.
     -- To avoid exploring them twice, we keep a cache of explored schedules.
-    createCacheST :: ST s (STRef s (Set ScheduleControl))
-    createCacheST = newSTRef $
+    cacheIO :: IO (IORef (Set ScheduleControl))
+    cacheIO = newIORef $
               -- we use opts here just to be sure the reference cannot be
               -- lifted out of exploreSimTrace
               if explorationScheduleBound opts >=0
                 then Set.empty
                 else error "exploreSimTrace: negative schedule bound"
 
-    cacheSizeST :: STRef s (Set ScheduleControl) -> ST s Int
-    cacheSizeST = fmap Set.size . readSTRef
+    cachedIO :: ScheduleControl -> IO Bool
+    cachedIO m = atomicModifyIORef' cache $ \set ->
+      (Set.insert m set, Set.member m set)
+
+    cacheSizeIO :: () -> IO Int
+    cacheSizeIO () = Set.size <$> readIORef cache
 
 
 -- |  Trace `SimTrace` to `stderr`.
 --
-traceDebugLog :: Int -> SimTrace a -> ST s ()
-traceDebugLog logLevel _trace | logLevel <= 0 = pure ()
-traceDebugLog 1 trace = Debug.traceM $ "Simulation trace with discovered schedules:\n"
-                                    ++ Trace.ppTrace (ppSimResult 0 0 0) (ppSimEvent 0 0 0) (ignoreRaces $ void `first` trace)
-traceDebugLog _ trace = Debug.traceM $ "Simulation trace with discovered schedules:\n"
-                                    ++ Trace.ppTrace (ppSimResult 0 0 0) (ppSimEvent 0 0 0) (void `first` trace)
```

Comment:
It would be nice to keep `traceDebugLog` in some form, it was used in `exploreSimTrace`.  Note that it's implemented with `Debug.traceM` so it can run in any monad, if we use `Debug.trace` instead it could be used in pure setting as well.